### PR TITLE
+ builder: emit implicit hash passed to a method call as kwargs

### PR DESCRIPTION
**Pull request by iliabylich**
Closes https://github.com/whitequark/parser/issues/761.

### CODE REVIEW

**Comment by iliabylich** [comment]
@marcandre Could you take a look, too, please? (Kind find you in a "Reviewers" dropdown)

File: lib/parser/builders/default.rb [hunk 1]
```diff
@@ -128,6 +128,54 @@ class << self
...
+      #       (int 42))))
+      # ```
+      #
+      # Note that `kwargs` node is just a replacement for `hash` argument,
+      # so if there's are multiple arguments (or a `kwsplat`) all of them
+      # are wrapped into `kwargs` instead of `hash`:
+      #
+      # ```
+      # (send nil :foo
+      #   (hash
```
Comment (marcandre) [inline lib/parser/builders/default.rb:164]: ^hash^kwargs

**Review by marcandre — COMMENTED** [review]
The implementation surprises me. It was too difficult to emit the right node type from the get go?

**Comment by iliabylich** [comment]
> It was too difficult to emit the right node type from the get go?

Yes and no, it's relatively simple. We rewrite either the last arg (if it's a hash) or the 2nd last (if it's a hash AND the last argument is a block). Rules are not trivial, and so the implementation takes 5 lines. The rest of the diff is about docs/tests.

(just in case) If you are talking about rewriting it in the `associate` method (that constructs a `hash` node) that's simply incorrect, not all hashes are kwargs. And there's no node for `parameters` (that'd be very helpful here)

**Comment by marcandre** [comment]
I meant that looking in `ruby30.y` there are 4 rules that call `associate`; I was expecting those in `primary` and `aref_args` to remain unchanged, and those in  `opt_call_args` and in `call_args` to change to `associate_kwargs` (or whatever you'd want to call them). Would that not work?

**Comment by iliabylich** [comment]
Yes, I had a similar idea, but there's a single exception - `indexasgn`:
```
             lhs: ...
                | primary_value tLBRACK2 opt_call_args rbracket
                    {
                      result = @builder.index_asgn(val[0], val[1], val[2], val[3])
                    }
```

Here `opt_call_args` can't be `kwargs` because this `foo[bar: 42] = baz` doesn't take kwargs at all.

Also by looking deeper at the grammar I've realized that it would affect `super(foo: 42)`, and we need to modify it too 😄 . I'll push an update

**Comment by iliabylich** [comment]
Or maybe it's safer to rewrite it at the very beginning and un-rewrite in indexasgn. Yes, it seems to be the only exception, all other branches end up doing `call_method`/`super`/`yield`

**Comment by iliabylich** [comment]
I give up on moving it to the grammar level. On 1.8 (huh) `aref_args` rule is a part of indexing (that does support kwargs) and regular array literals at the same time (that don't support it). Making 1.8 special would require us to rewrite a ton of tests to separate 1.8/SINCE_1_9 examples.

Added `kwargs` to `super` as mentioned above, going to merge it as is, feel free to propose better ideas.

**Pull request merged**